docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+- [Introduction to vectorization with Vector128 and Vector256](#introduction-to-vectorization-with-vector128-and-vector256)
+  * [Code structure](#code-structure)
+    + [Testing](#testing)
+    + [Benchmarking](#benchmarking)
+      - [Custom config](#custom-config)
+      - [Memory alignment](#memory-alignment)
+        * [Enforcing memory alignment](#enforcing-memory-alignment)
+        * [Memory randomization](#memory-randomization)
+  * [Loops](#loops)
+    + [Scalar remainder handling](#scalar-remainder-handling)
+    + [Vectorized remainder handling](#vectorized-remainder-handling)
+    + [Access violation testing](#access-violation-av-testing)
+  * [Loading and storing vectors](#loading-and-storing-vectors)
+    + [Loading](#loading)
+    + [Storing](#storing)
+    + [Casting](#casting)
+  * [Mindset](#mindset)
+    + [Edge cases](#edge-cases)
+    + [Scalar solution](#scalar-solution)
+    + [Vectorized solution](#vectorized-solution)
+  * [Toolchain](#toolchain)
+    + [Creation](#creation)
+    + [Bit operations](#bit-operations)
+    + [Equality](#equality)
+    + [Comparison](#comparison)
+    + [Math](#math)
+    + [Conversion](#conversion)
+    + [Widening and Narrowing](#widening-and-narrowing)
+    + [Shuffle](#shuffle)
+  * [Summary](#summary)
+    + [Best practices](#best-practices)
+
+TL;DR: Go to [Summary](#summary)
+
+# Introduction to vectorization with Vector128 and Vector256
+
+Vectorization is the art of converting an algorithm from operating on a single value at a time to operating on a set of values (vector). It can greatly improve performance at a cost of increased code complexity.
+
+In recent releases, .NET has introduced many new APIs for vectorization. The vast majority of them are hardware specific, so they require users to provide an implementation per processor architecture (x64 and/or arm64), with the option of using the most optimal instructions for hardware that is executing the code.
+
+.NET 7 introduced a set of new APIs for `Vector128<T>` and `Vector256<T>` for writing hardware-agnostic, cross platform vectorized code. The purpose of this document is to introduce you to the new APIs and provide a set of best practices.
+
+## Code structure
+
+`Vector128<T>` represents a 128-bit vector of type `T`. `T` is constrained to specific primitive types:
+
+* `byte` and `sbyte` (8 bits).
+* `short` and `ushort` (16 bits).
+* `int`, `uint` and `float` (32 bits).
+* `long`, `ulong` and `double` (64 bits).
+* `nint` and `unit` (32 or 64 bits, depending on the architecture)
```

Comment:
These two are new to .NET 7. They are not supported on .NET 3.1, 5, or 6.

.NET 8 is introducing a `Vector128<T>.IsSupported` that helps identify whether a given `T` will throw or not to help identify what works per runtime, including from generic contexts.